File: LFU.java
import java.util.*;
//Least freuqently used Cache aka LFU implementation in Java
class LFUCache {
    private final int capacity;
    private int minFrequency;
    private final Map<Integer, Integer> keyToValue;
    private final Map<Integer, Integer> keyToFrequency;
    private final Map<Integer, LinkedHashSet<Integer>> frequencyToKeys;

    public LFUCache(int capacity) {
        this.capacity = capacity;
        this.minFrequency = 0;
        this.keyToValue = new HashMap<>();
        this.keyToFrequency = new HashMap<>();
        this.frequencyToKeys = new HashMap<>();
    }

    public int get(int key) {
        if (!keyToValue.containsKey(key)) {
            return -1;
        }
        // Increase the frequency of the key
        int frequency = keyToFrequency.get(key);
        keyToFrequency.put(key, frequency + 1);

        // Remove the key from the current frequency list
        frequencyToKeys.get(frequency).remove(key);

        // If this was the last key in the frequency list and this was the min frequency, update minFrequency
        if (frequencyToKeys.get(frequency).isEmpty()) {
            frequencyToKeys.remove(frequency);
            if (frequency == minFrequency) {
                minFrequency++;
            }
        }

        // Add the key to the new frequency list
        frequencyToKeys.computeIfAbsent(frequency + 1, k -> new LinkedHashSet<>()).add(key);
        return keyToValue.get(key);
    }

    public void put(int key, int value) {
        if (capacity == 0) {
            return;
        }

        // If the key is already present, update the value and its frequency
        if (keyToValue.containsKey(key)) {
            keyToValue.put(key, value);
            get(key); // To increase the frequency
            return;
        }

        // If the cache is full, evict the least frequently used key
        if (keyToValue.size() >= capacity) {
            // Get the least frequently used key (minFrequency)
            int keyToEvict = frequencyToKeys.get(minFrequency).iterator().next();
            frequencyToKeys.get(minFrequency).remove(keyToEvict);

            // Remove the key if it's the last one with this frequency
            if (frequencyToKeys.get(minFrequency).isEmpty()) {
                frequencyToKeys.remove(minFrequency);
            }

            keyToValue.remove(keyToEvict);
            keyToFrequency.remove(keyToEvict);
        }

        // Add the new key-value pair
        keyToValue.put(key, value);
        keyToFrequency.put(key, 1);
        frequencyToKeys.computeIfAbsent(1, k -> new LinkedHashSet<>()).add(key);
        minFrequency = 1; // Since it's a new key, its frequency is 1, so minFrequency is reset to 1
    }

File: LFU.cc
/*
  Design a data structure that follows the constraints of a Least Frequently Used (LFU) cache.
  Implement the LFUCache class:
  - LFUCache(int capacity) Initialize the LFU cache with positive size capacity.
  - int get(int key) Return the value of the key if the key exists, otherwise return -1.
  - void put(int key, int value) Update the value of the key if the key exists. Otherwise,
    add the key-value pair to the cache. If the number of keys exceeds the capacity from this
    operation, evict the least frequently used key. For ties, the least recently used key among
    them is evicted.
  - The functions get and put must each run in O(1) average time complexity.
*/
#include <list>
#include <unordered_map>

class LFUCache {
private:
    int capacity;
    int minFreq;

    // key -> (value, freq)
    std::unordered_map<int, std::pair<int, int>> keyValueFreq;

    // freq -> list of keys (front = most recently accessed, back = eviction candidate)
    std::unordered_map<int, std::list<int>> freqList;

    // key -> iterator in freqList[freq]
    std::unordered_map<int, std::list<int>::iterator> keyIter;

    void incrementFreq(int key) {
        int freq = keyValueFreq[key].second;

        freqList[freq].erase(keyIter[key]);
        if (freqList[freq].empty()) {
            freqList.erase(freq);
            if (minFreq == freq)
                minFreq++;
        }

        freq++;
        keyValueFreq[key].second = freq;
        freqList[freq].push_front(key);
        keyIter[key] = freqList[freq].begin();
    }

public:
    LFUCache(int capacity) : capacity(capacity), minFreq(0) {}

    int get(int key) {
        if (keyValueFreq.find(key) == keyValueFreq.end())
            return -1;

        incrementFreq(key);
        return keyValueFreq[key].first;
    }

    void put(int key, int value) {
        if (capacity <= 0) return;

        // if already present, update value and promote frequency
        if (keyValueFreq.find(key) != keyValueFreq.end()) {
            keyValueFreq[key].first = value;
            incrementFreq(key);
            return;
        }

        // if at capacity, evict LFU (least freq + oldest among ties)
        if (static_cast<int>(keyValueFreq.size()) == capacity) {
            int oldKey = freqList[minFreq].back();
            freqList[minFreq].pop_back();

            if (freqList[minFreq].empty())
                freqList.erase(minFreq);

            keyValueFreq.erase(oldKey);
            keyIter.erase(oldKey);
        }

        // insert new key with frequency 1
        keyValueFreq[key] = {value, 1};
        freqList[1].push_front(key);
        keyIter[key] = freqList[1].begin();
        minFreq = 1;
    }
};
